File: windows/src/LanStash.Domain/Photos/SynologyPhotoModels.cs
namespace LanStash.Domain;

/// <summary>Photos 项目身份不能由文件路径、文件名或缩略图单元推导。</summary>
public enum SynologyPhotoSpace { Personal, Shared }
public sealed record SynologyPhotoIdentity(Guid ProfileId, SynologyPhotoSpace Space, long ItemId);
public sealed record SynologyPhotoThumbnail(long UnitId, string Revision);

public sealed record SynologyPhoto(
    SynologyPhotoIdentity Id,
    string Filename,
    long SizeBytes,
    DateTimeOffset TakenAt,
    DateTimeOffset IndexedAt,
    long FolderId,
    string MediaType)
{
    public SynologyPhotoThumbnail? Thumbnail { get; init; }
    public int? Width { get; init; }
    public int? Height { get; init; }
    public int? Orientation { get; init; }
    public string? Description { get; init; }
    public string? Camera { get; init; }
    public string? Lens { get; init; }
    public string? Aperture { get; init; }
    public string? ExposureTime { get; init; }
    public string? FocalLength { get; init; }
    public string? Iso { get; init; }
    public double? Duration { get; init; }
    public int? Rating { get; init; }
    public IReadOnlyList<string> Address { get; init; } = [];
    public double? Latitude { get; init; }
    public double? Longitude { get; init; }

    public bool IsSameDeletionTarget(SynologyPhoto other) =>
        Id == other.Id && Filename == other.Filename && SizeBytes == other.SizeBytes &&
        TakenAt == other.TakenAt && IndexedAt == other.IndexedAt &&
        FolderId == other.FolderId && MediaType == other.MediaType;
}

public sealed record SynologyPhotoAccess(IReadOnlyList<SynologyPhotoSpace> Spaces, string PackageVersion);
public sealed record SynologyPhotoDay(DateOnly Date, int Count);
public sealed record SynologyPhotoPage(IReadOnlyList<SynologyPhoto> Items, int Offset, int NextOffset, bool HasMore);
public sealed record SynologyPhotoCollection(long Id, string Name, long? ParentId = null, int? Count = null);
public enum SynologyPhotoCategory { Recent, People, Subjects, Locations, Tags, Videos }
public enum SynologyPhotoShareScope { WithMe, WithOthers, Requests }
public sealed record SynologyPhotoSharedEntry(string Id, string Title, long? AlbumId = null, Uri? Url = null);
public sealed record SynologyPhotoChoice(long Id, string Name);
public sealed record SynologyPhotoLocation(long Id, string Name, int Level, IReadOnlyList<SynologyPhotoLocation> Children);
public sealed record SynologyPhotoFocalRange(int Start, int End);
public sealed record SynologyPhotoFraction(int Num, int Den);
public sealed record SynologyPhotoExposureRange(SynologyPhotoFraction Start, SynologyPhotoFraction End);

public sealed record SynologyPhotoFilter
{
    public int? MediaType { get; init; }
    public long? StartTime { get; init; }
    public long? EndTime { get; init; }
    public long? PersonId { get; init; }
    public long? LocationId { get; init; }
    public long? TagId { get; init; }
    public int? Rating { get; init; }
    public long? CameraId { get; init; }
    public long? LensId { get; init; }
    public long? IsoId { get; init; }
    public long? ApertureId { get; init; }
    public SynologyPhotoFocalRange? FocalRange { get; init; }
    public SynologyPhotoExposureRange? ExposureRange { get; init; }
    public bool IsActive => this != new SynologyPhotoFilter();
}

public sealed record SynologyPhotoFilterOptions
{
    public IReadOnlyList<SynologyPhotoChoice> People { get; init; } = [];
    public IReadOnlyList<SynologyPhotoLocation> Locations { get; init; } = [];
    public IReadOnlyList<SynologyPhotoChoice> Tags { get; init; } = [];
    public IReadOnlyList<SynologyPhotoChoice> Cameras { get; init; } = [];
    public IReadOnlyList<SynologyPhotoChoice> Lenses { get; init; } = [];
    public IReadOnlyList<SynologyPhotoChoice> Iso { get; init; } = [];
    public IReadOnlyList<SynologyPhotoChoice> Apertures { get; init; } = [];
    public IReadOnlyList<SynologyPhotoFocalRange> FocalRanges { get; init; } = [];
    public IReadOnlyList<SynologyPhotoExposureRange> ExposureRanges { get; init; } = [];
}

public abstract record SynologyPhotoQuery
{
    public sealed record Timeline(long Start, long End) : SynologyPhotoQuery;
    public sealed record Search(string Keyword, long Start, long End) : SynologyPhotoQuery;
    public sealed record Filtered(SynologyPhotoFilter Filter, long Start, long End) : SynologyPhotoQuery;
    public sealed record Category(SynologyPhotoCategory Kind, long Id, long Start, long End) : SynologyPhotoQuery;
    public sealed record Folder(long Id) : SynologyPhotoQuery;
    public sealed record Album(long Id) : SynologyPhotoQuery;
    public sealed record Recent : SynologyPhotoQuery;
}

public enum SynologyPhotoFailure
{
    Unavailable, Permission, InvalidResponse, Media, LengthMismatch,
    DeletionUnverified, TargetChanged, DeleteDenied, DeleteFailed,
}
public sealed class SynologyPhotoException(SynologyPhotoFailure failure) : Exception($"photos.{failure}")
{
    public SynologyPhotoFailure Failure { get; } = failure;
}
public enum SynologyPhotoDeletionResult { Confirmed, PendingReview }

/// <summary>播放器只拿到有界读取能力；凭据、远端 URL 和具体 HTTP 路由留在传输层。</summary>
public interface IReadOnlyMediaSource : IDisposable
{
    long Length { get; }
    string ContentType { get; }
    Task<byte[]> ReadAsync(long offset, int count, CancellationToken cancellationToken = default);
}

public sealed record SynologyPhotoMediaRequest(
    ApiCapability Capability, string Method, IReadOnlyDictionary<string, string> Parameters);

public interface ISynologyPhotosProvider
{
    ISynologyPhotosRepository SynologyPhotos { get; }
}

/// <summary>内部 Photos 契约，不提供 File Station 降级或路径转换。</summary>
public interface ISynologyPhotosRepository
{
    Guid ProfileId { get; }
    bool CanDeleteOriginals => false;
    Task<SynologyPhotoAccess> AccessAsync(CancellationToken cancellationToken = default);
    Task<IReadOnlyList<SynologyPhotoDay>> TimelineAsync(SynologyPhotoQuery? query = null, CancellationToken cancellationToken = default);
    Task<SynologyPhotoPage> PhotosAsync(SynologyPhotoQuery query, int offset, int limit, CancellationToken cancellationToken = default);
    Task<SynologyPhotoCollection> RootFolderAsync(CancellationToken cancellationToken = default) => Unsupported<SynologyPhotoCollection>();
    Task<IReadOnlyList<SynologyPhotoCollection>> FoldersAsync(long parentId, int offset, int limit, CancellationToken cancellationToken = default) => Unsupported<IReadOnlyList<SynologyPhotoCollection>>();
    Task<IReadOnlyList<SynologyPhotoCollection>> AlbumsAsync(int offset, int limit, CancellationToken cancellationToken = default) => Unsupported<IReadOnlyList<SynologyPhotoCollection>>();
    Task<IReadOnlySet<SynologyPhotoCategory>> CategoriesAsync(CancellationToken cancellationToken = default) => Unsupported<IReadOnlySet<SynologyPhotoCategory>>();
    Task<IReadOnlyList<SynologyPhotoCollection>> CategoryItemsAsync(SynologyPhotoCategory category, int offset, int limit, CancellationToken cancellationToken = default) => Unsupported<IReadOnlyList<SynologyPhotoCollection>>();
    Task<IReadOnlyList<SynologyPhotoSharedEntry>> SharedEntriesAsync(SynologyPhotoShareScope scope, int offset, int limit, CancellationToken cancellationToken = default) => Unsupported<IReadOnlyList<SynologyPhotoSharedEntry>>();
    Task<SynologyPhotoFilterOptions> FilterOptionsAsync(CancellationToken cancellationToken = default) => Unsupported<SynologyPhotoFilterOptions>();
    Task<byte[]> ThumbnailAsync(SynologyPhoto photo, bool large = false, CancellationToken cancellationToken = default) => Unsupported<byte[]>();
    Task<SynologyPhoto> DetailsAsync(SynologyPhoto photo, CancellationToken cancellationToken = default) => Unsupported<SynologyPhoto>();
    Task<IReadOnlyMediaSource> VideoSourceAsync(SynologyPhoto photo, CancellationToken cancellationToken = default) => Unsupported<IReadOnlyMediaSource>();
    Task DownloadOriginalAsync(SynologyPhoto photo, string destination, IProgress<long>? progress = null, CancellationToken cancellationToken = default) => Unsupported<bool>();
    Task PrepareDeletionAsync(SynologyPhoto photo, CancellationToken cancellationToken = default) => Unsupported<bool>();
    Task<SynologyPhotoDeletionResult> DeleteAsync(SynologyPhoto photo, Guid operationId, CancellationToken cancellationToken = default) => Unsupported<SynologyPhotoDeletionResult>();
    Task<SynologyPhotoDeletionResult> ReviewDeletionAsync(SynologyPhoto photo, CancellationToken cancellationToken = default) => Unsupported<SynologyPhotoDeletionResult>();

    private static Task<T> Unsupported<T>() => Task.FromException<T>(new SynologyPhotoException(SynologyPhotoFailure.Unavailable));
}

File: android/app/src/main/java/io/github/qwertyuiop1995/dsmnativeclient/domain/SynologyPhotoModels.kt
package io.github.qwertyuiop1995.dsmnativeclient.domain

import java.io.File
import java.time.LocalDate
import java.time.YearMonth
import java.util.UUID

/** Photos 项目不是 File Station 文件；身份不从目录、文件名或缩略图单元推导。 */
enum class SynologyPhotoSpace { PERSONAL, SHARED }

data class SynologyPhotoId(val profileId: String, val space: SynologyPhotoSpace, val itemId: Long)
data class SynologyPhotoThumbnail(val unitId: Long, val revision: String)
data class SynologyPhoto(
    val id: SynologyPhotoId,
    val filename: String,
    val sizeBytes: Long,
    val takenAt: Double,
    val indexedAt: Double,
    val folderId: Long,
    val mediaType: String,
    val thumbnail: SynologyPhotoThumbnail? = null,
    val width: Int? = null,
    val height: Int? = null,
    val orientation: Int? = null,
    val description: String? = null,
    val camera: String? = null,
    val lens: String? = null,
    val aperture: String? = null,
    val exposureTime: String? = null,
    val focalLength: String? = null,
    val iso: String? = null,
    val duration: Double? = null,
    val rating: Int? = null,
    val address: List<String> = emptyList(),
    val latitude: Double? = null,
    val longitude: Double? = null,
) {
    fun sameDeletionTarget(other: SynologyPhoto): Boolean =
        id == other.id && filename == other.filename && sizeBytes == other.sizeBytes &&
            takenAt == other.takenAt && indexedAt == other.indexedAt &&
            folderId == other.folderId && mediaType == other.mediaType
}

data class SynologyPhotoPage(
    val items: List<SynologyPhoto>, val offset: Int, val nextOffset: Int, val hasMore: Boolean,
)
data class SynologyPhotoAccess(val spaces: List<SynologyPhotoSpace>, val packageVersion: String)
data class SynologyPhotoDay(val date: LocalDate, val count: Int) {
    val month: YearMonth get() = YearMonth.from(date)
}
data class SynologyPhotoCollection(
    val id: Long, val name: String, val parentId: Long? = null,
    val count: Int? = null, val thumbnail: SynologyPhotoThumbnail? = null,
)
enum class SynologyPhotoCategory(val serverId: String, val parameter: String?) {
    RECENT("recently_added", null), PEOPLE("person", "person_id"),
    SUBJECTS("concept", "concept_id"), LOCATIONS("geocoding", "geocoding_id"),
    TAGS("general_tag", "general_tag_id"), VIDEOS("video", null),
}
enum class SynologyPhotoShareScope { WITH_ME, WITH_OTHERS, REQUESTS }
data class SynologyPhotoSharedEntry(
    val id: String, val title: String, val albumId: Long? = null, val url: String? = null,
)

data class SynologyPhotoChoice(val id: Long, val name: String)
data class SynologyPhotoLocation(
    val id: Long, val name: String, val level: Int,
    val children: List<SynologyPhotoLocation> = emptyList(),
)
data class SynologyPhotoFocalRange(val start: Int, val end: Int)
data class SynologyPhotoFraction(val num: Int, val den: Int)
data class SynologyPhotoExposureRange(val start: SynologyPhotoFraction, val end: SynologyPhotoFraction)

data class SynologyPhotoFilter(
    val mediaType: Int? = null,
    val startTime: Long? = null,
    val endTime: Long? = null,
    val personId: Long? = null,
    val locationId: Long? = null,
    val tagId: Long? = null,
    val rating: Int? = null,
    val cameraId: Long? = null,
    val lensId: Long? = null,
    val isoId: Long? = null,
    val apertureId: Long? = null,
    val focalRange: SynologyPhotoFocalRange? = null,
    val exposureRange: SynologyPhotoExposureRange? = null,
) {
    val isActive: Boolean get() = this != SynologyPhotoFilter()
}

data class SynologyPhotoFilterOptions(
    val people: List<SynologyPhotoChoice> = emptyList(),
    val locations: List<SynologyPhotoLocation> = emptyList(),
    val tags: List<SynologyPhotoChoice> = emptyList(),
    val cameras: List<SynologyPhotoChoice> = emptyList(),
    val lenses: List<SynologyPhotoChoice> = emptyList(),
    val iso: List<SynologyPhotoChoice> = emptyList(),
    val apertures: List<SynologyPhotoChoice> = emptyList(),
    val focalRanges: List<SynologyPhotoFocalRange> = emptyList(),
    val exposureRanges: List<SynologyPhotoExposureRange> = emptyList(),
)

sealed interface SynologyPhotoQuery {
    data class Timeline(val start: Long, val end: Long) : SynologyPhotoQuery
    data class Search(val keyword: String, val start: Long, val end: Long) : SynologyPhotoQuery
    data class Filtered(val filter: SynologyPhotoFilter, val start: Long, val end: Long) : SynologyPhotoQuery
    data class Category(val category: SynologyPhotoCategory, val id: Long, val start: Long, val end: Long) : SynologyPhotoQuery
    data class Folder(val id: Long) : SynologyPhotoQuery
    data class Album(val id: Long) : SynologyPhotoQuery
    data object Recent : SynologyPhotoQuery
}

enum class SynologyPhotoFailureKind {
    UNAVAILABLE, PERMISSION, INVALID_RESPONSE, MEDIA, LENGTH_MISMATCH,
    DELETION_UNVERIFIED, TARGET_CHANGED, DELETE_DENIED, DELETE_FAILED,
}
class SynologyPhotoFailure(val kind: SynologyPhotoFailureKind, cause: Throwable? = null) :
    RuntimeException(kind.name, cause)

enum class SynologyPhotoDeletionResult { CONFIRMED, PENDING_REVIEW }

/** 所有方法使用已发现且版本匹配的 Photos 接口；不回退文件系统扫描。 */
interface SynologyPhotosServing {
    val profileId: String
    val canDeleteOriginals: Boolean get() = false
    suspend fun access(): SynologyPhotoAccess
    suspend fun timeline(space: SynologyPhotoSpace, query: SynologyPhotoQuery? = null): List<SynologyPhotoDay>
    suspend fun photos(space: SynologyPhotoSpace, query: SynologyPhotoQuery, offset: Int, limit: Int): SynologyPhotoPage
    suspend fun rootFolder(space: SynologyPhotoSpace): SynologyPhotoCollection = unsupported()
    suspend fun folders(space: SynologyPhotoSpace, parentId: Long, offset: Int, limit: Int): List<SynologyPhotoCollection> = unsupported()
    suspend fun albums(offset: Int, limit: Int): List<SynologyPhotoCollection> = unsupported()
    suspend fun categories(): Set<SynologyPhotoCategory> = unsupported()
    suspend fun categoryItems(category: SynologyPhotoCategory, offset: Int, limit: Int): List<SynologyPhotoCollection> = unsupported()
    suspend fun sharedEntries(scope: SynologyPhotoShareScope, offset: Int, limit: Int): List<SynologyPhotoSharedEntry> = unsupported()
    suspend fun filterOptions(): SynologyPhotoFilterOptions = unsupported()
    suspend fun thumbnail(photo: SynologyPhoto, large: Boolean = false): ByteArray = unsupported()
    suspend fun details(photo: SynologyPhoto): SynologyPhoto = unsupported()
    suspend fun videoSource(photo: SynologyPhoto): RandomAccessMediaSource = unsupported()
    suspend fun downloadOriginal(photo: SynologyPhoto, destination: File, progress: (Long, Long) -> Unit = { _, _ -> }): Unit = unsupported()
    suspend fun prepareDeletion(photo: SynologyPhoto): Unit = unsupported()
    suspend fun deletePhoto(photo: SynologyPhoto, operationId: UUID): SynologyPhotoDeletionResult = unsupported()
    suspend fun reviewDeletion(photo: SynologyPhoto): SynologyPhotoDeletionResult = unsupported()

    private fun unsupported(): Nothing = throw SynologyPhotoFailure(SynologyPhotoFailureKind.UNAVAILABLE)
}
